golangci-lint: fix goimports & add staticcheck

olivierlemasle opened the pull request:
This PR brings 3 commits related to golangci-lint:

1. Fix `goimports` configuration (`.golangci.yml` had an indentation issue) and re-order imports;
2. Update golangci-lint:
   - Bump golangci-lint to its latest version (`1.50.1`)
   - Remove linters `deadcode`, `structcheck` and `varcheck`, deprecated by golangci-lint since version `1.49.0` (because they were unmaintained)
   - Remove linter `rowserrcheck`, which is disabled by golangci-lint because it's not compatible with Go generics, and which is not useful for metrics-server as it only checks errors in `sql.Rows`
   - Add linter `staticcheck`
3. Fix `staticcheck` errors

olivierlemasle commented on pkg/api/node.go:
```diff
@@ -137,11 +137,9 @@ func (m *nodeMetrics) ConvertToTable(ctx context.Context, object runtime.Object,
 	switch t := object.(type) {
 	case *metrics.NodeMetrics:
 		table.ResourceVersion = t.ResourceVersion
-		table.SelfLink = t.SelfLink
```
staticcheck:
> table.SelfLink is deprecated: selfLink is a legacy read-only field that is no longer populated by the system. +optional

serathius replied:
We cannot remove it willy nilly just because it's legacy. We need to be backward compatible, which this linter seems not understand. Deprecated doesn't mean "can be removed at will", it means "it will not be supported further on, so don't create new usage".

I would prefer to keep the old code to avoid breaking old clients that might rely on it.

olivierlemasle commented on pkg/scraper/client/resource/client.go:
```diff
@@ -112,12 +111,12 @@ func (kc *kubeletClient) getMetrics(ctx context.Context, url, nodeName string) (
 	buf.Reset()
 	_, err = io.Copy(buf, response.Body)
 	if err != nil {
-		kc.buffers.Put(b)
```
staticcheck:
> [SA6002](https://staticcheck.io/docs/checks#SA6002): argument should be pointer-like to avoid allocations (staticcheck)

olivierlemasle replied:
:warning: I'm not totally sure of my fix.

serathius replied:
Don't agree with the check statement, allocations are not always bad. And don't recommend changing this code without allocation benchmarking.

olivierlemasle commented on pkg/scraper/client/resource/decode_test.go:
```diff
@@ -244,7 +246,7 @@ container_start_time_seconds{container="metrics-server",namespace="kubernetes-da
 }
 
 func Fuzz_decodeBatchPrometheusFormat(f *testing.F) {
-	testSeedsFloat64 := []float64{0, -10000, 10000, 0.5, -0.000000001, -0.0, 1e100, -1e100}
```
staticcheck:
> in Go, the floating-point literal '-0.0' is the same as '0.0', it does not produce a negative zero

olivierlemasle commented on test/e2e_test.go:
```diff
@@ -432,15 +431,15 @@ func proxyRequestToPod(config *rest.Config, namespace, podname, scheme string, p
 }
 
 func setupForwarding(config *rest.Config, namespace, podname string, port int) (cancel func(), err error) {
-	hostIP := strings.TrimLeft(config.Host, "https://")
```
`strings.TrimLeft` returns a slice of the string s with all leading Unicode code points contained in cutset removed.

`TrimPrefix` was intended here.

olivierlemasle commented:
/retest

yangjunmyfm192085 reviewed (commented):
/cc

yangjunmyfm192085 commented on Makefile:
```diff
@@ -14,9 +14,12 @@ BUILD_DATE:=$(shell date -u +'%Y-%m-%dT%H:%M:%SZ')
+
 # Computed variables
 # ------------------
-HAS_GOLANGCI:=$(shell which golangci-lint)
```
Why is modification needed here?
Is the original way inappropriate?

olivierlemasle replied:
Well, replacing `which golangci-lint` by `golangci-lint version --format=short` is not strictly required (I can revert if you prefer); I used it to make `make verify-lint` use a **fixed / determined** version of golangci-lint. Currently, if you have already locally an old version of `golangci-lint` on your path, `make golangci` will not update it, which means you will not use the same version of golangci-lint as CI. That can lead to discrepancies.

Furthermore:
- `HAS_GOLANGCI` was defined twice, with the same value (line 19 and line 222)
- line 226, `$(go env GOPATH)` does not work (it should be `$(shell go env GOPATH)`, or simply `$(GOPATH)`, because `GOPATH` variable is defined line 20.
- Thus, on my local environment (and I guess, for most developer environments), `make golangci` tries to install golangci-lint in `/bin`, which is denied without sudo privileges. In CI, installing in `/bin` works. However, in CI, `$(GOPATH)/bin` is not in binary path; that's why I replaced `golangci-lint` by `$(GOPATH)/bin/golangci-lint` in the other make rules.

Note: I used the same makefile rules when I added golangci-lint to prometheus-adapter in https://github.com/kubernetes-sigs/prometheus-adapter/pull/540.

yangjunmyfm192085 replied:
OK. I have no other questions. Maybe we need in-depth reviews from others
/cc @serathius @dgrisonnet

dgrisonnet commented on pkg/scraper/client/resource/client.go:
```diff
@@ -68,7 +67,8 @@ func newClient(c *http.Client, resolver utils.NodeAddressResolver, defaultPort i
 		buffers: sync.Pool{
 			New: func() interface{} {
-				return make([]byte, 10e3)
+				b := make([]byte, 10e3)
```
```suggestion
				b := new(bytes.Buffer)
				b.Grow(10e3)
```

dgrisonnet commented on pkg/scraper/client/resource/client.go:
```diff
@@ -107,17 +107,18 @@ func (kc *kubeletClient) getMetrics(ctx context.Context, url, nodeName string) (
 		return nil, fmt.Errorf("request failed, status: %q", response.Status)
 	}
-	b := kc.buffers.Get().([]byte)
+	bRef := kc.buffers.Get().(*[]byte)
+	b := *bRef
 	buf := bytes.NewBuffer(b)
```
These 3 lines could be simplified with the following if we change the pool initialization function:

```
b := bufPool.Get().(*bytes.Buffer)
```


With that we would be able to avoid having to create a new buffer and could just reuse the one from the pool.

olivierlemasle replied:
Good idea, I've updated the PR.

dgrisonnet reviewed (approved):
/lgtm

/assign @serathius

yangjunmyfm192085 reviewed (commented):
/retest

yangjunmyfm192085 reviewed (approved):
Can you squash the commits?

olivierlemasle commented:
@yangjunmyfm192085 Sure, I've rebased and squashed.

yangjunmyfm192085 reviewed (commented):
/lgtm

olivierlemasle commented:
/retest

yangjunmyfm192085 commented:
/retest

serathius commented on Makefile:
```diff
@@ -213,17 +216,17 @@ endif
-ifndef HAS_GOLANGCI
-	curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(go env GOPATH)/bin v1.48.0
+ifneq ($(HAS_GOLANGCI_VERSION), $(GOLANGCI_VERSION))
+	curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(GOPATH)/bin v$(GOLANGCI_VERSION)
```
Replacing `$(go env GOPATH)/bin` with `$(GOPATH)/bin` can lead to unexpected results. Difference between them is that `$GOPATH` is en environment variable that can be empty and leading to aforementioned path to be resolved to `/bin`. This is not wanted as will lead to user confusion when they get permission error for accessing `/bin`.

 `$(go env GOPATH)` will not only return `$GOPATH` environment value, but also will provide a fallback based on localtion of go binary.

olivierlemasle replied:
@serathius Actually, in the context of this Makefile:
- `$GOPATH` is **not** an environment variable, but a Make variable, whose value is the output of `go env GOPATH` (initialized [line 23](https://github.com/kubernetes-sigs/metrics-server/blob/0556b85536db201fe82558a99cff679596c8b161/Makefile#L23)). It is therefore **not empty**.
- On the contrary, `$(go env GOPATH)` resolves to `""` ($ is for Make variables); `$$(go env GOPATH)` would work ($$ is for shell variables).

Here is a simple Makefile to show it better than words:
```make
GOPATH:=$(shell go env GOPATH)

.PHONY: var
var:
	echo $(GOPATH)
	# output: my local GOPATH path

.PHONY: one
one:
	echo $(go env GOPATH)
	# output: nothing

.PHONY: two
two:
	echo $$(go env GOPATH)
	# output: my local GOPATH path
```

serathius replied:
Missed that. Thanks for detailed explanation.

olivierlemasle commented:
@serathius Thanks for the review.
I've reverted the changes related to the deprecated field and the allocations in the buffer, and I've added `//nolint` comments for these cases.

serathius commented:
/approve
/lgtm
